src/MgvOfferMaking.sol

```diff
@@ -326,11 +332,17 @@ contract MgvOfferMaking is MgvHasOffers {
         ofp.local =
           dislodgeOffer(offerList, ofp.olKey.tickScale, ofp.oldOffer, ofp.local, cachedLocalTick, shouldUpdateBranch);
         // If !shouldUpdateBranch, then ofp.local.level0 and ofp.local.level1 reflect the removed tick's branch post-removal, so one cannot infer the tick by reading those fields. If shouldUpdateBranch, then the new tick must be inferred from the new info in local.
+        // FIXME check if shouldUpdateBranch was true but became false inside dislodgeOffer
         if (shouldUpdateBranch) {
-          cachedLocalTick = ofp.local.bestTick();
+          // force control flow through gas-saving path if retraction emptied the offer list
+          if (ofp.local.level0().isEmpty()) {
+            cachedLocalTick = insertionTick;
+          } else {
+            cachedLocalTick = ofp.local.bestTick();
+          }
         }
       }
-      if (insertionTick.strictlyBetter(cachedLocalTick)) {
+      if (!cachedLocalTick.strictlyBetter(insertionTick)) {
```

Comment:
why also update if they are equal?

Reply:
We encore « empty book » with « insertion tick = current tick ». I tried maintaining the original condition and add « OR level0 is empty » and I got very mysterious gas increases from a couple hundred to thousand ds in some tests. I let it be but it’s a mystery open for any interested detective.